org.lflang/src/org/lflang/validation/LFValidator.java

```diff
@@ -1199,6 +1206,122 @@ public void checkVarRef(VarRef varRef) {
         }
     }
 
+
+    /**
+     * Specification of the structure of an annotation.
+     */
+    static class AttributeSpec {
+
+        private static final String VALUE_ATTR = "value";
+        private final Map<String, AttrParamSpec> paramSpecByName;
+
+
+        public AttributeSpec(List<AttrParamSpec> params) {
+
+            paramSpecByName = params.stream().collect(Collectors.toMap(it -> it.name, it -> it));
+
+        }
+
+        /**
+         * Check that the attribute conforms to this spec. The
+         * attr has the correct name.
+         */
+        public void check(LFValidator validator, Attribute attr) {
```

Comment:
Javadoc is missing for these new methods.

Reply:
Javadoc added in the new commits.